emcc.py

```diff
@@ -1578,6 +1578,14 @@ def default_setting(name, new_default):
       default_setting('ERROR_ON_UNDEFINED_SYMBOLS', 0)
       default_setting('WARN_ON_UNDEFINED_SYMBOLS', 0)
 
+    if shared.Settings.EXCEPTION_CATCHING_ALLOWED:
+      if 'DISABLE_EXCEPTION_CATCHING' in settings_key_changes:
```

Comment:
How is this line different from
```py
if shared.Settings.DISABLE_EXCEPTION_CATCHING:
```
? If they are different, should we also check if both `shared.Settings.DISABLE_EXCEPTION_CATCHING` and `shared.Settings.EXCEPTION_CATCHING_ALLOWE` are set?

Reply:
This is set if the user specified it explicitly on the command line (regardless of if the value matches the default).  

Doing that here allows me to show a warning (or error) if the user ever passed both these options at the same time.

Reply:
Then shouldn't we check if both of those two flags are specified in
1. One in `Settings` and the other in the command line (and vice versa)
2. Both in `Settings`
3. Both in the command line

Maybe I could be mistaken.

Reply:
In theory yes, but I don't think we are careful about local modifications to `settings.js`. That is, users should use the commandline. Changes to `settings.js` by an emscripten developer would assume they know the risk of changing internals.

Reply:
If users are supposed to use the command line, then we should check the case when both of them are in `settings_key_change` at least?

Also most of these checking lines in this function directly refer to not `settings_key_change` but `shared.Settings` directly, which is weird if users are supposed use the command line.

Reply:
In general yes, although I think this code covers the cases we care about.

I don't think (2) is much of an issue.  I don't think folks should be directly modifying `settings.js`, at least not without also also modifying code like this that is somewhat aware of the default.   So I think its fair to say that if `EXCEPTION_CATCHING_ALLOWED` is set then its the user that set it.

That means if these two conditions are both two they specified both on the command line.  That is exactly the case we care about erroring/warning for.   I agree we could possibly be more robust here and not make assumptions about the default values.

I added a comment which should hopefully make it a little more clear what is going on.

Reply:
Sorry, what I was asking is, if we only care about the command line,
why are we not checking `settings_key_change` for `EXCEPTION_CATCHING_ALLOWED` as we do for `DISABLE_EXCEPTION_CATCHING`? We are checking `settings_key_changes` for one and checking `shared.Settings` for the other. This is the part I don't understand.

Also, if we only care about command lines, why are all other lines in this function directly checking `shared.Settings` and now `settings_key_changes`?

Reply:
> If users are supposed to use the command line, then we should check the case when both of them are in `settings_key_change` at least?
> 
> Also most of these checking lines in this function directly refer to not `settings_key_change` but `shared.Settings` directly, which is weird if users are supposed use the command line.

Technically you are correct yes, we should also check for `EXCEPTION_CATCHING_ALLOW` in `settings_key_changes`.

Maybe this latest version is more explicit.

Reply:
Thanks, but the problem here is my lack of knowledge of the code around here ;( Sorry. Can someone answer why we check `shared.Settings` directly in the code within this function, if we expect the users to use only the command line? I asked why we were not using `settings_key_changes` for both settings, but other code around this line all check `shared.Settings` directly, so I'm wondering why this should be, or if this should be, different, and how all this works.

When are we merging `settings_key_changes` to `shared.Settings`? Before running this code or after? If after, checking `settings.Changes` only makes sense, but if so you wouldn't have checked `settings_key_changes` in the first place...

Sorry for many questions. If taking offline is easier it's fine too.

Reply:
`shared.Settings` reflects that actual current settings which is the sum of:

a) the defaults
b) any user command line settings
c) any changes made by the running of the code here (which updates a lot of settings).

In almost all cases `shared.Settings` is the only thing we care about because it represent the actual current setup.

`settings_key_changes` is for cases were we actually care about what the user specified on the commandline as opposed to (a) or (c).    Its mostly used for diagnostics stuff like this.    For example in this case we need to use it so we can tell the different between the default `=0` and the command line `=0`.

I will make a couple of PRs to clean this up a bit and give `settings_key_changes` a better name :)